README.md
# SRCclock: decodifica e riproduce il Segnale Rai Codificato

## Italiano

Questo programma decodifica/riproduce il segnale radiio RAI SRC.
Il segnale SRC (Segnale Rai Codificato) è generato negli studi di Via Asiago a Roma dalla RAI.
Qualsiasi sintonizzatore AM o FM incorporato oppure attaccato al computer può essere utilizzato per ricevere e decodificare il segnale SRC con questo programma.
Alternativamente, è possibile leggere l'audio RAW da un file o riprodurre il segnale per motivi di test.

## English

This program decodes/plays the SRC RAI radio signal.
The SRC signal (Segnale orario Rai Codificato) is generated by RAI in the studios located in Via Asiago in the city of Rome.
Any AM/FM radio tuner incorporated or attached to the computer can be used to receive and to decode the SRC through this program.
Alternatively, it is possible to read the raw audio data from file or to reproduce the signal for test purposes.


Changelog.md
# Changelog del progetto

* 03/03/2019 **(v 1.1)**
	* Traduzione in italiano del codice
	* Traduzione in italiano della documentazione
	* Tutti i file di testo sono ora scritti con la notazione markdown
	* Rimozione di ogni riferimento al nome dell'INRiM dato che dal 1º gennaio 2017 l'azienda pubblica radiotelevisiva si autoproduce il segnale dagli studi di Via Asiago a Roma
	* Cambiamento della modalità di generazione del file e rimozione dei file non più necessari
	* Spostamento di alcuni file
	* Miglioramento della leggibilità del codice

* 01/04/2014 **(v 1.0)**
	* Bug fixes
	* Documentation

* 05/03/2014 **(v 0.4)**
	* Command line arguments
	* Bug fixes

* 03/12/2013 **(v 0.3)**
	* Window Decision System feature

* 08/11/2013 **(v 0.2)**
	* *Unknown changes*


SRCclock/source/crw.h
#ifndef CRW_H
#define CRW_H

#include <string>
#include <fstream>
#include <pulse/simple.h>
#include <pulse/error.h>
//#include <pulse/pulseaudio.h>

class Crw {
    int		INstate;			// Stato dello stream in input
    int		OUTstate;			// Stato dello stream in output

    std::fstream	fd;			// Descrizione del file per l'input e l'output su file
    pa_simple*	sound_stream;			// Descrizione dello stream di PulseAudio

    int		sound_error;			// Variabile che identifica l'ultimo errore nello stream del suono

public:

  /**
   * @brief Leggi i dati dallo stream di input che è stato precedentemente aperto
   *
   * @param Buffer dei dati da leggere
   * @param samples Numero di campioni da leggere. Nessun controllo viene eseguito sulle dimensioni dell'array di buffer
   * @param size Dimensione in bytes di ciascun elemento del buffer. Di default è settato a 1
   * @return <int> Numero di campioni letti
   **/
   
  virtual int readBuffer(void* buffer, int samples, int size = 1);		// Leggi un buffer dallo stream di input aperto

  /**
   * @brief Scrive dei dati allo stream di input che è stato precedentemente aperto
   *
   * @param buffer Buffer di dati da scrivere
   * @param samples Numero di campioni da scrivere. Nessun controllo viene eseguito sulle dimensioni dell'array di buffer
   * @param size Dimensione in bytes di ciascun elemento del buffer. Di default è settato a 1
   * @return <int> Dimensione in bytes di ciascun elemento del buffer. Di default è settato a 1
   **/

  virtual int writeBuffer(const void* buffer, int samples, int size = 1);	// Scrivi un buffer a uno stream aperto di output

    /**
     * @brief Apre un semplice stream di PulseAudio in playback
     *
     * @param fc Frequenza di campionamento
     * @param canali 1 = mono; 2 = stereo
     * @param SampFormat Formato di campionamento
     * @param device Nome del dispositivo di PulseAudio
     * @param appName Nome dell'applicazione
     * @return <bool> Ritorna True se tutto è OK
     **/

    virtual bool open_soundStream_output(int fc, int channels, const char* device = 0, pa_sample_format SampFormat = PA_SAMPLE_U8, const char* appName = 0);

    /**
     * @brief Apre un semplice stream di PulseAudio in registrazione
     *
     * @param fc Frequenza di campionamento
     * @param canali 1 = mono; 2 = stereo
     * @param SampFormat Formato di campionamento
     * @param device Nome del dispositivo di PulseAudio
     * @param appName Nome dell'applicazione
     * @return <bool> Ritorna True se tutto è OK
     **/

    virtual bool open_soundStream_input(int fc, int channels, const char* device = 0, pa_sample_format SampFormat = PA_SAMPLE_U8, const char* appName = 0);


    virtual bool open_file_output(const char* fileNAme);	// Apre lo stream di output su file
    virtual bool open_file_input(const char* fileNAme);		// Apre lo stream di input su file

    virtual void close_input_stream();				// Chiude lo stream di input
    virtual void close_output_stream();				// Chiude lo stream di output
    virtual void close_all();					// Chiude tutti gli streams



    std::string get_sound_error() const { return pa_strerror(sound_error); }	// Ritorna la stringa del precedente errore di PulseAudio


    int get_INstate() const { return INstate; }    // Ritorna il valore della variable INstate per lo stato dell'input
    int get_OUTstate() const { return OUTstate; }	// Ritorna il valore della variable OUTstate per lo stato dell'output

    Crw();
    virtual ~Crw() { close_all(); }
};

#endif // CRW_H


SRCclock/source/csrc.h
#ifndef CSRC_H
#define CSRC_H

#include <cmath>
#include <ctime>
#include <vector>
#include <string>
#include <cstdlib>
#include <chrono>
#include "crw.h"
#include "clog.h"

using std::vector;
using std::string;

class Csrc : public Crw {

  const int F0;		// frequency of tone 0
  const int F1;		// frequency of tone 1
  const int Fsync;
  const double Ts;	// time of each symbol

  int year, month, day, wday, hour, min, sec;
  int leap_second, change_time;
  bool dst;

  int msec;		// milliseconds. Used to compensate the error on the syncronisation due to post processing of the samples

  vector<int> src_vector;

  int timeout;			// timeout in sec
  int soundChannels;		// mono, stereo

  int sample_frequency;
  double decision_threshold;	// decision threshold in dB
  bool adaptive_decision_threshold;
  int  window_length;
  double snr_level;

  int verbose_level;		// verbose level for debugging messages

  bool decoded;		// decode status
  bool running;
  bool do_sync;

  int error;
  bool streamON;	// the stream has been established

  Clog lout, lerr;	// logs for output and error messages

  std::chrono::high_resolution_clock::time_point sys_clock;

public:
    Csrc();		/**< Default constructor */
    virtual ~Csrc();	/**< Descructor */

    /**
     * @brief Open the output stream on the Sound server. The SRC signal will be played there.
     *
     * @param fc Sampling frequency
     * @param channels Channels: 1 = mono; 2 = stereo. Default: mono
     * @param device Name of the sound device. Leaving it NULL will choose the default sound card. The sound server is the one used as default.
     * @param SampFormat Format of the samples. The SRC class expects float samples in the continues range [-1;1]. Please, don't change this parameter unless sure. On Intel machines the endianness is 'Little Endian'.
     * @param appName Name of the application using the sound server. By default "SRC".
     * @return bool
     */
    bool open_soundStream_output(int fc, int channels = 1, const char* device = NULL, pa_sample_format SampFormat = PA_SAMPLE_FLOAT32LE, const char* appName = "SRC");


    /**
     * @brief Open the input stream on the Sound server. The SRC signal will be played there.
     *
     * @param fc Sampling frequency
     * @param channels Channels: 1 = mono; 2 = stereo. Default: mono
     * @param device Name of the sound device. Leaving it NULL will choose the default sound card. The sound server is the one used as default.
     * @param SampFormat Format of the samples. The SRC class expects float samples in the continues range [-1;1]. Please, don't change this parameter unless sure. On Intel machines the endianness is 'Little Endian'.
     * @param appName Name of the application using the sound server. By default "SRC".
     * @return bool
     */
    bool open_soundStream_input(int fc, int channels = 1, const char* device = NULL, pa_sample_format SampFormat = PA_SAMPLE_FLOAT32LE,const char* appName = "SRC");

    /**
     * @brief Open the output stream on File. The default sampling frequency is 8000 Hz on one (mono) channel
     *
     * @param fileNAme Name of the file
     * @return bool
     */
    bool open_file_output(const char* fileNAme);


    /**
     * @brief Open the input stream on File. The default sampling frequency is 8000 Hz on one (mono) channel
     *
     * @param fileNAme Name of the file
     * @return bool
     */
    bool open_file_input(const char* fileNAme);


    /**
     * @brief Open the output stream on file.
     *
     * @param fileNAme Name of the file
     * @param fc Sampling frequency
     * @param channels Channels: 1 = mono; 2 = stereo.
     * @return bool
     */
    bool open_file_output(const char* fileNAme, int fc, int channels);



    /**
     * @brief Open the input stream on file.
     *
     * @param fileNAme Name of the file
     * @param fc Sampling frequency
     * @param channels Channels: 1 = mono; 2 = stereo.
     * @return bool
     */
    bool open_file_input(const char* fileNAme, int fc, int channels);



    /**
     * @brief Write raw data on the output stream if it is open. The number of samplers written are returned.
     *
     * @param buffer Pointer to sampling data
     * @param samples Number of samples
     * @return int number of samples written
     */
    int writeBuffer(const float* buffer, int samples);


   /**
     * @brief Read raw data on the input stream if it is open. The number of samplers read are returned.
     *
     * @param buffer Pointer to sampling data
     * @param samples Number of samples
     * @return int number of samples read
     */
    int readBuffer(float* buffer, int samples);




    void close_input_stream();		/**< Close input stream */
    void close_output_stream();		/**< Close output stream */
    void close_all();			/**< Close all active streams */


    /**
     * @brief Play the SRC signal on the output stream.
     *
     * @param power Power level of the signal in dB (maximum level is 0 dB)
     * @param initial_delay flag that indicates whether to put a random number of samples on the stream before playing the SRC. Duration is less than 1 second.
     * @param random_theta Flag that indicates wheter to add a random \f$\theta\f$ to the signal
     * @param noise_sigma RMS of the Additive White Gaussian Noise added to the main signal
     */
    void play(double power = -3, bool initial_delay = false, bool random_theta = false, double noise_sigma = 0.0);

    bool warnings() const { return (change_time != 7) || (leap_second != 0); }	/**< Return true if there is any warning issued. */

    bool operator==(const Csrc& other) const;	/**< Equal operator between objects of class Csrc */
    bool operator<(const Csrc& other) const;	/**< Minor operator between objects of class Csrc */
    Csrc& operator=(const Csrc& other);		/**< Operator = for class Csrc */


    vector<int> get_src_vector() const { return src_vector; }	/**< Return the binary vector in the for of a vector<int> */
    int* get_src_vector_int() const;	/**< Return the binary vector in the for of a int* */


    bool OK() const { return decoded; }	/**< If the decoding was successful is returned TRUE */

    bool P1() const;	/**< Check on the parity 1 (first block) */
    bool P2() const;	/**< Check on the parity 2 (first block) */
    bool PA() const;	/**< Check on the parity of the second block */

    bool ID1() const;	/**< Return true if the first identification ID1 is valid */
    bool ID2() const;	/**< Return true if the second identification ID2 is valid */


    int OR() const { return hour; }		/**< Return the hour */
    int MI() const { return min; }		/**< Return the minute */
    bool OE() const { return dst; }		/**< Return the flag of summer time */
    int ME() const { return month; }		/**< Return the month */
    int GM() const { return day; }		/**< Return the day */
    int GS() const { return wday; }		/**< Return the day of the week */
    int AN() const { return year; }		/**< Return the year */
    int SE() const { return change_time; }	/**< Return the warning of change time */
    int SI() const { return leap_second; }	/**< Return the warning of leapsecond */


    bool valid_date() const;	/**< check if the format of the date is valid */
    bool check(int bits = 48);	/**< check the first bits of the received sequence */



    void set_today();		/**< The internal variables are set to the current date and time */
    void reset();		/**< Reset all the variables the default and close all the stream */
    void stop();		/**< Any action on streams are stopped. */
    bool streamOK() const { return streamON; }	/**< Return the state of the stream. A 0 is returned for invalid input or output streams. */



    int	get_timeout() const { return timeout; }	/**< Return the decoding timeout in seconds */
    void set_timeout(int seconds);		/**< Set the decoding timeout in seconds */




    bool decode();		/**< decode the SRC signal. Return TRUE is decoding was successful */

    bool sincronized() const { return decoded && (error == 0); }	/**< Return TRUE if the SRC has been decoded and sincronized */


    /**
     * @brief Set the length in term on time symbols for the Window Decision System
     *
     * The Window Decision System is an algorithm used by the class in order to adapt the value of the decision threshold
     * to the noise level of the channel. The average of the power associated to both signal frequencies is calculated over the last
     * Wlength time symbols and the decision threshold is set snr dB over this level. This way the decision threshold is kept updated
     * and adapted to the channel noise level.
     * The WDS algorithm is used by the Csrc class in order to select the right value of any decision thresholds and it
     * is alternative with respect the static decision. It is assumed here that if the window length is positive then a
     * dynamic evaluation of the channel is required. By the contrary, any zero or negative value of the window length
     * implies the use of the static threshold, whose value can be changed with the function set_decision_threshold().
     *
     * @param Wlength dimension of the window in time symbols. If Wlength <= 0 then no WDS adaptation will be made
     * @param snr value indicating the increasing of the decision threshold in dB with respect the noise average
     * @return length of the window in time symbols
     */
    int setWDS(int Wlength, double snr = 12.0);





    double set_decision_threshold(double dB);		/**< Set the value in dB of the decision threshold */
    inline double get_decision_threshold() const;	/**< Return the value of the decision threshold in dB */




    void yes_sync() { do_sync = true; }	/**< Syncronisation ON. In case of play mode, the syncronisation tones are played. In decoding mode, the syncronisation tones are used to syncronize with the minute of the SRC. */
    void no_sync() { do_sync = false; }	/**< Skip syncronisation */
    bool get_sync() const { return do_sync; }	/**< Return the syncronisation mode */





    int internalError() const { return error; }	/**< Return the internal error level. */



    /**
      * @brief Set the verbose level. Higher is the value and more detailed is the debugging output.
      *
      * @param level Level of the verbose. 1 = normal verbose; 2 = debugging level 1; 3 = debugging level 2; less than 1: no verbose.
      */
    void set_verbose(int level) { verbose_level = level; }
    int get_verbose() const { return verbose_level; }	/**< Return the verbose level */


    static bool leapyear(int y);	/**< Return true if the year has 366 days */


    friend std::ostream& operator<<(std::ostream& os, const Csrc& src);	/**< Output of the binary string of the SRC */
    friend std::istream& operator>>(std::istream& is, Csrc& src);	/**< Input of the binary string of the SRC */

    string dateSTD() const;	/**< Return a string of the date in the standard format defined in RFC 2822 */
    string dateISO() const;	/**< Return a string of the date in the standard format ISO 8601 */
    struct tm get_date_tm() const;	/**< Return a tm struct of the date/time information */

    /**
      * @brief Return the string of the SRC date/time with format RFC2822 ora ISO8601
      *
      * @param iso8601 This variable should be TRUE to have the date/time in the format ISO8601, FALSE for the format defined in RFC2822
      */
    string dateSTR(bool iso8601 = false) const;



    void logOnFile(const char* fileName);	/**< The log are redirect on external file */
    void errorLogOnFile(const char* fileName);	/**< Errors log are redirect on external file */
    void logOnSTDOUT();				/**< Set the default STDOUT stream for logs */



    /**
      * @brief Set internal information of date and time plus the warning flags.
      *
      * @param MI Minutes [0, 60]
      * @param OR Hours [0, 23]
      * @param GM Day of the month [1, 31]
      * @param GS Day of the week [1, 7] (1 = Monday)
      * @param ME Month [1, 12]
      * @param AN Year  (0 = 2000, 1 = 2001, etc...)
      * @param OE flag of summer time
      * @param SE Warning of change time [0, 7] (7 = no warnings, 6 = change time in 6 days, 5 = change time in 5 days, ...)
      * @param SI Warning of leap second at the end of the month (-1, 0, +1)
      * @return True if the date is correct
      */
    bool set(int MI, int OR, int GM, int GS, int ME, int AN, bool OE = false, int SE = 7, int SI = 0);


    /**
      * @brief Set internal information of date and time through the struct tm. No warning flags for change time or leapsecond are set
      *
      * @param ttmm Is a struct tm with the information related to the date and time to be set.
      * @return True if the date is correct
      */
    bool set(const struct tm& ttmm);


    /**
      * @brief Set internal information of date and time. The date/time are passed through a string and processed by function strptime().
      *
      * @param strDate Pointer to a string of char containing the date
      * @param format Pointer to a string of char containing the format of the date passed in the parameter strDate. Please, refer to the documentation of standard function strptime() to set the correct format string.
      * @return True if the date is correct. If any kind of error is encountered during the process of the date string, the internal variables are set to the current date and time.
      */
    bool set(const char* strDate, const char* format);




    /**
      * @brief Set the warning flags.
      *
      * @param SE Warning of change time [0, 7] (7 = no warnings, 6 = change time in 6 days, 5 = change time in 5 days, ...)
      * @param SI Warning of leap second at the end of the month (-1, 0, +1)
      * @return True if the warnings are correct
      */
    bool setWarnings(int SE, int SI);


    bool setOE(bool newDST);	/**< Function to set the flag of the OE (dst) of the SRC bit-string */


    /**
      * @brief Return the microseconds elapsed since last reading from the input stream.
      *
      * This function is useful to perform fine sincronisation of the system clock. Each time some samples are read
      * from the input stream the internal timing variable is reset and the computational delay can be calculated
      * through this function. Essentially, once the SRC signal has been received the system clock is set to the
      * current date/time decoded plus the milliseconds and microseconds elapsed since last reading. The precision
      * can be of the orther of few microseconds depending on the machine's hardware.
      *
      */
    long microsecDelay() const;

    int getMilliseconds() const { return msec;}	/**< Return the number of milliseconds after last reference second */


    int number_of_RP() const;	/**< Return the expected number of syncronisation tones RP for the current date/time */


// private functions:
private:

  double randn(double mu, double sigma);
  int parity(int beg, int end) const;
  void bin_convert(int val, int offset, int length);	// converts from an integer value to bit
  int deconvert(int offset, int length);
  inline void checkSample(float& s);
  inline void stereo_encode(float* p, int k);
  void encode();	// build the src_vector


  double goertzel(int frequency, const float* data, int samples) const;		// calculates the power associated to a given frequency with the Goertzel algorithm
  int read_buffer(float* b, int total_size, int& bytes2read, int& extra);	// modified function for reading the buffer
  int tuning(int freq, float* buffer, int& extra, int N, int DELTA, int STEP, double& p);
  static void reset_buffer(float* b, int size, float value = 0.0);
  static string itos(int value, int length = 1, int base = 10, char fill = '0', bool force_sign = false);
  void add_minute();
  inline double max(double a, double b) const;
  inline void reset_src_vector();
};

#endif // CSRC_H


SRCclock/source/main.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <getopt.h>
#include <sys/time.h>
#include "csrc.h"

using std::cout;
using std::cerr;


// Struttura utilizzata per raccogliere le opzioni da riga di comando
struct SRCoption {
	int SRCaction;
	bool random_samples, random_theta,  do_sync, binary, iso, sys_sync;
	int verb, chdate, leap, fc, channels, timeout, wds, repeat, dst;
	double th, power, noise, snr_level;
	long delay;
	char *soundDev, *fo, *logfile, *setDate;
};




void print_help(const char* filename)
{
  cout<<"Utilizzo:\t" <<filename <<" --decode [OPZIONI]\n"
	<<"oppure\t" <<filename <<" --play [OPZIONI]\n"
	<<"\nLista delle opzioni disponibili:\n"
	<<"  -d, --decode\t\tDecodifica il segnale SRC\n"
	<<"  -y, --system-sync\tSincronizza l'ora di sistema con il segnale SRC. Necessita\n\t\t\t dei privilegi di super user\n"
	<<"  -N, --snr=SNR_LEVEL\tLivello di rilevamento SNR sul rumore in dB.\n\t\t\tDi default è SNR_LEVEL=5dB sopra il livello di rumore\n"
	<<"  -W, --window=LENGTH\tWindow Decision System. Imposta la lunghezza della finestra\n\t\t\tin time symbols. Di default è LENGTH=50 simboli\n"
	<<"  -D, --delay=DELAY\tdelay della sincronizzazione in microsecondi\n"
	<<"  -T, --timeout=TIMEOUT\tImposta il timeout per la decodifica in secondi\n"
	<<"  -t, --threshold=TH\tImposta la soglia di decisione statica in dB(default -35 dB)\n"
	<<'\n'
	<<"  -p, --play\t\tRiproduci il segnale SRC\n"
	<<"  -k, --rand-theta\tDa un theta random in riproduzione\n"
	<<"  -o, --rand-samples\tAggiungi dei campioni random (< 1 sec) prima di riprodurre il segnale\n"
	<<"  -a, --power=POWER\tLivello di potenza del segnale in dB in riproduzione\n"
	<<"  -n, --noise=SIGMA\tRiproduce il segnale SRC con un rumore espresso come il RMS\n"
	<<"  -S, --set-date=DATE\tImposta la data/ora del segnale SRC nel formato <hh:mm DD/MM/YYYY>\n"
	<<"  -e, --dst-on\t\tForza il flag per l'ora solare (OE) ad essere 1\n"
	<<"  -E, --dst-off\t\tForza il flag per l'ora solare (OE) ad essere 0\n"
	<<"  -C, --change-time=SE\tImposta i giorni di avviso SRC per il cambiamento dell'ora legale/solare [0-7]\n\t\t\t(7 = senza avviso, default)\n"
	<<"  -l, --leap-second=SI\tImposta l'avviso SRC del secondo intercalare (+1|-1)\n\t\t\talla fine del mese\n"
	<<'\n'
	<<"  -s, --sync\t\tSincronizzazione con gli impulsi sonori da 1kHz\n\n"
	<<"  -r, --rate=FREQ\tImposta la frequenza di campionamento (per l'input/output su file\n\t\t\tdi default è 8 kHz)\n"
	<<"  -m, --mono\t\tImposta l'audio in mono (1 canale). Di default è stereo\n\t\t\t(2 canali)\n"
	<<"  -M, --stero\t\tImposta l'audio in stereo (2 canali). Questa è\n\t\t\tl'opzione di default\n"
	<<"  -f, --file=FILE\tSeleziona il file sorgente/destinazione (l'impostazione predefinita è il server audio)\n"
	<<"  -c, --card=DEV\tSpecifica un diverso dispositivo audio\n"
	<<"  -v, --debug=LEVEL\tLivello verbose (default = 1)\n"
	<<"  -I, --iso\t\tStampa la data/ora nel formato ISO 8601\n\t\t\t(predefinito: formato RFC2822)\n"
	<<"  -b, --binary\t\tStampa la rappresentazione binaria del segnale SRC\n"
	<<"  -R, --repeat=TIMES\tNumero di ripetizioni di decodifica (predefinito = 1.\n\t\t\tImposta 0 per ripetizioni illimitate)\n"
	<<"  -L, --logfile=LOG\tReindirizzare le uscite al registro file\n"
	<<"  -w, --warranty\tDettagli sulla garanzia\n"
	<<"  -V, --version\t\tVersione del programma\n"
	<<"  -h, --help\t\tStampa nuovamente questo aiuto e poi esce\n\n";
}



void print_warranty()
{
  cout <<"Questo programma viene fornito senza alcuna garanzia\n"
       <<"Questo è software libero, e sei invitato a modificarlo, eseguirlo e ridistribuirlo\n"
			 <<"Questo è software è sotto licenza GPL v3\n";
}



int main(int argc, char **argv) {

  Csrc SRC;
  SRCoption options;
  const char* VERSION = "1.1";
  int error = 0;
  int choice;

  bool openState;	// used for error control


  // default values:
  options.random_samples = false;
  options.random_theta = false;
  options.do_sync = false;
  options.binary = false;
  options.iso = false;
  options.sys_sync = false;
  options.dst = 0;		// default, set by system date
  options.verb = 1;		// normal verbose level
  options.chdate = 7;		// no change date
  options.leap = 0;
  options.fc = 8000;		// 8 kHz
  options.channels = 2;		// stereo, 2 channels
  options.th = -35;
  options.power = -6.0;
  options.noise = 0.0;
  options.timeout = 0;		// uses default
  options.wds = 50;
  options.snr_level = 5.0;	// 5 dB SNR default

  options.fo = NULL;
  options.soundDev = NULL;	// default sound device
  options.logfile = NULL;
  options.setDate = NULL;
  options.repeat = 1;
  options.SRCaction = 0;
  options.delay = 1;



  if(argc == 1) {
    cerr <<"Errore: è richiesta almeno un'opzione\n\n";
    print_warranty();
    print_help(argv[0]);
    return 1;
  }

  int longindex = 0;	// variable used for the getopt_long function
  static struct option long_options[] = {
		{"decode",       no_argument,       NULL, 'd'},
		{"play",         no_argument,       NULL, 'p'},
		{"system-sync",  no_argument,       NULL, 'y'},
		{"delay",        required_argument, NULL, 'D'},
		{"rand-theta",   no_argument,       NULL, 'k'},
		{"rand-samples", no_argument,       NULL, 'o'},
		{"set-date",     required_argument, NULL, 'S'},
		{"dst-on",       no_argument,       NULL, 'e'},
		{"dst-off",      no_argument,       NULL, 'E'},
		{"power",        required_argument, NULL, 'a'},
		{"sync",         no_argument,       NULL, 's'},
		{"file",         required_argument, NULL, 'f'},
		{"debug",        required_argument, NULL, 'v'},
		{"threshold",    required_argument, NULL, 't'},
		{"noise",        required_argument, NULL, 'n'},
		{"snr",          required_argument, NULL, 'N'},
		{"window",       required_argument, NULL, 'W'},
		{"change-time",  required_argument, NULL, 'C'},
		{"leap-second",  required_argument, NULL, 'l'},
		{"rate",         required_argument, NULL, 'r'},
		{"mono",         no_argument,       NULL, 'm'},
		{"stereo",       no_argument,       NULL, 'M'},
		{"card",         required_argument, NULL, 'c'},
		{"iso",          no_argument,       NULL, 'I'},
		{"binary",       no_argument,       NULL, 'b'},
		{"timeout",      required_argument, NULL, 'T'},
		{"logfile",      required_argument, NULL, 'L'},
		{"repeat",       required_argument, NULL, 'R'},
		{"warranty",     no_argument,       NULL, 'w'},
		{"version",      no_argument,       NULL, 'V'},
		{"help",         no_argument,       NULL, 'h'},
		{0, 0, 0, 0}};


  while((choice = getopt_long(argc, argv, "dypkoeEa:sf:v:t:N:W:n:C:l:c:mMr:D:R:T:L:S:bIhVw", long_options, &longindex)) != -1) {
    switch (choice) {
      case 'd': options.SRCaction |= 1;		// SRCaction=1	=>	DECODE!
		break;
      case 'y': options.sys_sync = true;
		options.SRCaction |= 1;
		break;
      case 'p': options.SRCaction |= 2;		// SRCaction=2	=>	PLAY!
		break;
      case 'k': options.random_theta = true;
		options.SRCaction |= 2;
		break;
      case 'o': options.random_samples = true;
		options.SRCaction |= 2;
		break;
      case 'S': options.setDate = optarg;
		options.SRCaction |= 2;
		break;
      case 'e': options.dst = 1;
		options.SRCaction |= 2;
		break;
      case 'E': options.dst = 2;
		options.SRCaction |= 2;
		break;
      case 'a': options.power = atof(optarg);
		options.SRCaction |= 2;
		break;
      case 's': options.do_sync = true;
                break;
      case 'f': options.fo = optarg;
		break;
      case 'v': options.verb = atoi(optarg);
		break;
      case 't':	options.th = atof(optarg);
		options.SRCaction |= 1;
		break;
      case 'N': options.snr_level = atof(optarg);
		options.SRCaction |= 1;
		break;
      case 'W': options.wds = atoi(optarg);
		options.SRCaction |= 1;
		break;
      case 'n': options.noise = atof(optarg);
		options.SRCaction |= 2;
		break;
      case 'C': options.chdate = atoi(optarg);
		if((options.chdate < 0) || (options.chdate > 7)) {
		  cerr <<"Errore: errore di inserimento della data di modifica! Il parametro deve essere compreso nell'intervallo [0-7]\n";
		  options.chdate = 7;	// no changes
		}
		options.SRCaction |= 2;
		break;
      case 'l': options.leap = atoi(optarg);
		if((options.leap != 0) && (options.leap != -1) && (options.leap != 1)) {
		  cerr <<"Errore: il secondo intercalare deve essere 1 o -1 (nessun secondo intercalare => 0)\n";
		  options.leap = 0;
		}
		options.SRCaction |= 2;
		break;
      case 'r': options.fc = atoi(optarg);
		if((options.fc < 8000) || (options.fc > 48000)) {
		  cerr <<"Errore: la frequenza di campionamento deve essere compresa tra 8000 e 48000 Hz\nDefault: 8000 Hz\n";
		  options.fc = 8000;
		}
		break;
      case 'm': options.channels = 1;
		break;
      case 'M': options.channels = 2;
		break;
      case 'c': options.soundDev = optarg;
		break;
      case 'D': options.delay = atol(optarg);
		options.SRCaction |= 1;
		break;
      case 'I': options.iso = true;
		break;
      case 'b': options.binary = true;
		break;
      case 'T': options.timeout = atoi(optarg);
		if(options.timeout < 2) {
		  cerr <<"Errore: il timeout non può essere inferiore a 2 secondi. Impostazione predefinita -> 300 secondi\n";
		  options.timeout = 300;
		}
		options.SRCaction |= 1;
		break;
      case 'L': options.logfile = optarg;
		break;
      case 'R': options.repeat = atoi(optarg);
		break;
      case 'w': print_warranty();
		break;
      case 'V': cout <<"Versione: " <<VERSION <<'\n';
		break;
      case '?': // unrecognized argument
		break;
      case 'h': print_help(argv[0]);
		return 0;
      default:	cerr <<"Errore: argomento sbagliato!\n";
		print_help(argv[0]);
		return 1;
    }
  }


  if(options.verb >= 4) {
    cout <<"Opzione passata:\n\n"
	 <<"decodeSRC: " <<options.SRCaction <<'\n'
	 <<"Window Decision System length: " <<options.wds <<'\n'
	 <<"random_theta: " <<options.random_theta <<'\n'
	 <<"random_samples: " <<options.random_samples <<'\n'
	 <<"System sincronisation: " <<options.sys_sync <<'\n'
	 <<"Power level: " <<options.power <<" dB\n"
	 <<"Do sync: " <<options.do_sync <<'\n'
	 <<"Verbose level: " <<options.verb <<'\n'
	 <<"WDS: " <<options.wds <<'\n'
	 <<"SNR level: " <<options.snr_level <<'\n'
	 <<"Noise RMS: " <<options.noise <<'\n'
	 <<"Change date: " <<options.chdate <<'\n'
	 <<"Leap second: " <<options.leap <<'\n'
	 <<"Sampling frequency: " <<options.fc <<" Hz\n"
	 <<"Channels: " <<options.channels <<'\n'
	 <<"Repeat: " <<options.repeat <<'\n'
	 <<"Timeout: " <<options.timeout <<'\n'
	 <<"Sync delay: " <<options.delay <<'\n';

    if(options.fo) cout <<"FileStream: " <<options.fo <<'\n';
    if(options.soundDev) cout <<"Sound device: " <<options.soundDev <<'\n';
    if(options.logfile) cout <<"Logfile: " <<options.logfile <<'\n';
    if(options.setDate) cout <<"Set date: " <<options.setDate <<'\n';
    cout <<'\n';
  }

  SRC.set_verbose(options.verb);

  if(options.logfile) SRC.logOnFile(options.logfile);
  else SRC.logOnSTDOUT();

  if(options.do_sync) SRC.yes_sync();
  else SRC.no_sync();

  if((options.SRCaction != 1) && (options.SRCaction != 2)) {
    cerr <<"Errore: è necessario specificare l'opzione -d o -p\n\n";
    print_help(argv[0]);
    return 1;
  }


  do {	// repetition loop
    if(options.SRCaction == 1) {
      if(options.fo) {
        openState = SRC.open_file_input(options.fo, options.fc, options.channels);
        if(!openState) {
	  cerr <<"Errore: impossibile aprire il file di input del flusso " <<options.fo <<'\n';
	  return 1;
        }
      }
      else {
        openState = SRC.open_soundStream_input(options.fc, options.channels, options.soundDev);
        if(!openState) {
	  cerr <<"Errore: impossibile aprire il flusso di input audio. " <<SRC.get_sound_error() <<'\n';
	  return 1;
        }
      }


      SRC.set_decision_threshold(options.th);
      SRC.setWDS(options.wds, options.snr_level);
      SRC.set_timeout(options.timeout);
      SRC.decode();

      error = SRC.internalError();
      if(options.sys_sync && SRC.sincronized()) {		// syncronisation of the system clock
        struct timeval t;
        struct timezone tz;	// timezone
        tm ttmm;

        tz.tz_minuteswest = 60*(1 + int(SRC.OE()));
//      tz.tz_dsttime = DST_MET;		// central europe dst time constant. For old systems....

        ttmm = SRC.get_date_tm();		// 1) return the tm structure of the date/time
        t.tv_sec  = mktime(&ttmm);	// 2) convert the current time to the number of seconds since the "epoc"
																		// 3) set the number of microseonds. These are calculated taking into account the delay since last reading plus an uncertainty value
        t.tv_usec = SRC.getMilliseconds()*1000l + SRC.microsecDelay() + options.delay;

        error = settimeofday(&t, &tz);
        if(error == 0) cout <<"Ora di sistema sincronizzata!\n" <<SRC.dateSTD() <<'\n';
        else cerr <<"EE: Impossibile sincronizzare l'ora di sistema\n";

      }
    }
    else if(options.SRCaction == 2) {
      if(options.fo)
        SRC.open_file_output(options.fo, options.fc, options.channels);
      else
        SRC.open_soundStream_output(options.fc, options.channels, options.soundDev);

      if(SRC.get_OUTstate() == 0) {
        cerr <<"Errore nell'apertura del flusso di output!\n";
        return 1;
      }

      if(options.setDate) SRC.set(options.setDate, "%H:%M %d/%m/%Y");
      switch (options.dst) {
        case 1:	SRC.setOE(true);
		break;
        case 2: SRC.setOE(false);
		break;
        default: break;
      }

      if((options.chdate != 7) || (options.leap != 0))
        SRC.setWarnings(options.chdate, options.leap);

      SRC.play(options.power, options.random_samples, options.random_theta, options.noise);
      if(options.verb >= 1) {
        cout <<SRC.dateSTR(options.iso) <<'\n';
        if(options.binary) cout <<SRC <<'\n';
      }
    }

    if(options.verb >= 0) {
      if(options.SRCaction == 1) {
        if(SRC.OK()) {
	  cout <<SRC.dateSTR(options.iso) <<'\n';
	  if(SRC.warnings() && (options.verb >= 1)) {
	    cout <<"---------------\nAVVISI SRC:\n";
	    if(SRC.SE() != 7) cout <<"Cambio dell'ora tra " <<SRC.SE() <<" giorni\n";
	    if(SRC.SI()) cout <<"Secondo intercalare alla fine del mese: " <<SRC.SI() <<" secondi\n";
	    cout <<"---------------\n";
	  }
          if(options.binary) cout <<SRC <<'\n';
        }
        else {
  	  cerr <<"Errore di decodifica\n";
        }
      }
    }

    SRC.close_all();
    options.repeat--;
  } while(options.repeat != 0);


  return error;
}
